File: plugin/src/main/java/de/s42/tradecycle/strategy/KeyboardStrategy.java
package de.s42.tradecycle.strategy;

import de.s42.tradecycle.event.VillagerCycleTradeEvent;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.entity.Villager;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.ClickType;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.MerchantInventory;

public class KeyboardStrategy implements Listener {

    @EventHandler
    public void onInventoryClickEvent(InventoryClickEvent event) {
        if (event.isCancelled()) {
            return;
        }

        if (event.getClick() != ClickType.SWAP_OFFHAND) {
            return;
        }

        Player player = (Player) event.getWhoClicked();
        Inventory topInventory = player.getOpenInventory().getTopInventory();
        if (!(topInventory instanceof MerchantInventory merchantInventory)) {
            return;
        }

        if (!(merchantInventory.getHolder() instanceof Villager villager)) {
            return;
        }

        event.setCancelled(true);
        Bukkit.getPluginManager().callEvent(new VillagerCycleTradeEvent(player, villager));
    }
}

File: core/src/main/kotlin/tr/s42/core/strategy/KeyboardStrategy.kt
package net.cutecraft.core.strategy

import net.cutecraft.core.event.VillagerCycleTradeEvent
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.entity.Villager
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.ClickType
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.MerchantInventory

class KeyboardStrategy : Listener {

    @EventHandler
    fun onInventoryClickEvent(event: InventoryClickEvent) {
        if (event.isCancelled) {
            return
        }

        if (event.click != ClickType.SWAP_OFFHAND) {
            return
        }

        val player = event.whoClicked as Player
        val topInventory: Inventory = player.openInventory.topInventory
        if (topInventory !is MerchantInventory) {
            return
        }

        val holder = topInventory.holder
        if (holder !is Villager) {
            return
        }

        event.isCancelled = true
        Bukkit.getPluginManager().callEvent(VillagerCycleTradeEvent(player, holder))
    }

}